### ISSUE TITLE
9.8 | RFC Release Candidate 1 ROADMAP

### ISSUE THREAD

**Issue by llaville** [overtrue/phplint#248]
This report superceeded the #239.

## 2026-08-18

- [x] The new `Metadata` subsystem is features completed (see #245)
- [x] The `OutputManager` extension is features completed (see #226)
- [x] The `DiagnoseManager` extension is features completed (see #238)
- [x] The `ProfileManager` extension is features completed (see #243)
- [x] The `ProgressManager` extension is features completed (see #222)
- [x] The Extension Architecture is now stable and features completed (see #221)
- [x] The `ArgumentResolver` is features completed (see #241)
- [x] The `lint` command is now modernized and use the invokable syntax (see #240)
 
What are still missing before to publish the first release candidate ?

- [ ] Fix all remaining test cases 
- [ ] Upgrade all documentation to reflect new features of version 9.8
- [ ] Make Docker integration tests
- [ ] Make [MegaLinter](https://github.com/oxsecurity/megalinter/) integration tests
- [ ] Upgrade the `CacheManager` extension (following tests if necessary)

**Comment by llaville** [comment]
## 2026-08-31

Version `9.8.x-dev (ec06833)`

### New features since 2026-08-18 milestone

1. By default, even if env var `PLINT_MODE` is not defined, the value is equal to `off`. That means that either when you run : 

```shell
bin/phplint
```

or 

```shell
PLINT_MODE=off bin/phplint
```

Then the default Symfony Console `list` command is invoked

2. When env var `PLINT_MODE` is set to `legacy`, we have a pretty behaviour than version 9.7.x : no need to specify the `lint` command, it's the default invocation

> [!WARNING]
> Beware that you have to enter a filename or a directory (the current working dir is not the default, as in verison 9.7)
  
3. The new env var `PLINT_LOGGER` allows to specify which class (PSR-3 compatible logger) you want to use to log events, command, plugins, arguments resolved and whatever PHPLint is able to generate.

By default, (even if not defined), the `PLINT_LOGGER` is set to the new `\Overtrue\PHPLint\Console\ConsoleLogger` component. An enhanced version of the basic Symfony Console ConsoleLogger. 
See the `bootstrap.php` example to handle new styles/colors.

To understand the differences, run following commands to compare :  

> [!NOTE]
> I've explicitly removed the progress bar and diagnostic output to limit results of this demo, but you can keep it (default values are `--progress auto` and `--diagnostic auto`)

```shell
PLINT_MODE=legacy bin/phplint src/ --progress never --diagnostic never -vvv
```

```shell
PLINT_MODE=legacy bin/phplint src/ --progress never --diagnostic never --bootstrap bootstrap.php -vvv
```

You can also compare results between new ConsoleLogger and Symfony version

```shell
PLINT_MODE=legacy PLINT_LOGGER="Symfony\Component\Console\Logger\ConsoleLogger" bin/phplint src/ --progress never --diagnostic never --bootstrap bootstrap.php -vvv
```

4. An `AbstractManager` class was introduced by commit https://github.com/overtrue/phplint/commit/31d37b526dde1ef1fd21b4e763b86fe7431c3cd6 to log all events with the new `ConsoleLogger` component.

5. All extensions (plugins) concept are now handled in a different way : by contract and implementation.

When env var `PLINT_ENV` is set to `dev` (development environment), default behaviour, all extensions are allowed : see the `lint --help` panel
You can change this behaviour by setting the new env vars `PLINT_ALLOW_PLUGINS` and `PLINT_DEFAULT_PLUGINS` to match your needs. 
 
### API changes since 2026-08-18 milestone

1. The `ConfigResolver` is now deprecated on all uses, but keep only that have a pseudo API backward compatible copy with previous version 9.7.x

2. The default binary launcher `phplint.php` was modified to allow to change color styles easily with the bootstrapping feature : see `bootstrap.php` file

**Comment by llaville** [comment]
## 2026-09-01

Version `9.8.x-dev (c330795)`

The color styles for the new `ConsoleLogger` component was integrated by default. You don't need anymore to use the `bootstrap.php` example file to do so. 
If you want to change colors to match your needs, you still can do it with the bootstrapping feature (`bootstrap.php` is a good example to do it).

> [!WARNING]
> Since yesterday, I've added a new section style called `SectionEnum::DEPENDENCY` to handle features that depends of optional packages like CPU auto detection, Symfony DotEnv, or CI auto detection. The color schema is the one defined by `SectionEnum::DEFAULT`

Some minor issues detected was fixed since previous milestone version (yesterday).

**Comment by llaville** [comment]
## 2026-09-03

Version `9.8.x-dev (f8fc31d)`

The `DiagnoseManager` extension/plugin reached is final step before RC-1
A documentation is available online at https://github.com/overtrue/phplint/blob/9.8/docs/extensions/diagnoseManager.md

**Comment by llaville** [comment]
## 2026-09-07

Version `9.8.x-dev (5abaa78)`

PHPLint 9.8 supports multiple console invocation modes always controlled by the `PLINT_MODE` environment variable.

Following three modes, were added to existing `off` and `legacy`

- `develop`
- `diagnose`
- `profile`

Console usage documentation was updated accordingly.

Inspired by the famous Xdebug 3 mode feature !

**Comment by llaville** [comment]
For user who don't want to spend lot of time to discover new features of the PHPLint 9.8 version, I've [added a FAQ](https://github.com/overtrue/phplint/blob/4bc3434b563b0351f0f9a8e29184b19a0a38cf67/docs/FAQ.md) to learn all what you need at a glance !

**Comment by llaville** [comment]
## 2026-09-17

Version `9.8.x-dev (7606eb8)`

- Fixed the PHAR distribution compilation

    ```shell
    BOX_REQUIREMENT_CHECKER=0 box-4.7.phar compile [<-c box.json>|<-c box.json.dist>]
    ```

    While we still use the https://github.com/box-project/box to build the PHAR copies, we have now important differences to know:
  
    - the BOX configuration file `box.json.dist` is used to build a PHAR distribution **without optional packages**, but have a Manifest (`--manifest` flag)
    - the BOX configuration file `box.json` is used to build a PHAR distribution **with all optional packages**, but does not have a Manifest


- The Source Code has now reached a stable statement milestone !

**Comment by llaville** [comment]
## 2026-09-20

Version `9.8.x-dev (78e1ee3)`

- [x] Fix all remaining test cases (allow me to find a nasty bug in our `ArgumentResolver` component)
- [ ] Upgrade all documentation to reflect new features of version 9.8
- [x] Make Docker integration tests
- [ ] Make [MegaLinter](https://github.com/oxsecurity/megalinter/) integration tests
- [x] The `CacheManager` extension is features completed (https://github.com/overtrue/phplint/issues/223)

Additional tasks since first roadmap version, one month ago !

- [x] PHAR distribution is ready (see differences between `box.json` and `box.json.dist` config files on my previous comment)
- [x] job limit and parallelization (see https://github.com/overtrue/phplint/issues/246)

As you see, there are only two tasks to do. We are pretty closed than ever to have our First Release Candidate Version published in next days.

**Comment by llaville** [comment]
## 2026-09-21

Version `9.8.x-dev (a724fbc)`

This is the last beta version before RC-1.

- MegaLinter integration was already initialized (even if it's not perfect currently)
- MegaLinter tests told me that  a profiler diagnostic was missing.  Now added with commit https://github.com/overtrue/phplint/commit/a724fbc3350f761f77ad5621ca918871e315395c
- I'll prepare all missing documentation, I think it's useful

And then, at least, a first release candidate

**Comment by llaville** [comment]
First release candidate will be published tomorrow.

- code source won't changed
- documentation may be still updated : but I hope you will here to contribute to make it better, once you'll have read it

**Comment by llaville** [comment]
While preparing the first RC, i've reverted https://github.com/overtrue/phplint/issues/248#issuecomment-5709096182 PHAR build behavior by commit https://github.com/overtrue/phplint/commit/04cff20977247652ba1895e98be7ddc38d36b6ac

- both `box.json` and `box.json.dist` config files include now all optional dependencies
- difference is with BOX Manifest:  included with `box.json.dist` , and excluded with `box.json`

**Comment by llaville** [comment]
RC-1 delayed to one more day

**Comment by llaville** [comment]
Version [9.8.0-rc.1](https://github.com/overtrue/phplint/releases/tag/9.8.0-rc.1) has just been published
